### PR TITLE
Initial Policy Framework

### PR DESCRIPTION
**Pull request by mtarking**
Initial Policy Framework that supports:
- bring your own policy
- use case driven policy over time

### CODE REVIEW

File: plugins/action/common/prepare_plugins/prep_107_policy.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        policy_switches = model_data['vxlan']['policy']['switches']
+        for switch in policy_switches:
+            if any(sw['name'] == switch['name'] for sw in topology_switches):
+                found_switch = next((item for item in topology_switches if item["name"] == switch['name']))
+                if found_switch.get('management').get('management_ipv4_address'):
+                    switch['name'] = found_switch['management']['management_ipv4_address']
+                elif found_switch.get('management').get('management_ipv6_address'):
+                    switch['name'] = found_switch['management']['management_ipv6_address']
+
+        self.kwargs['results']['model_extended'] = model_data
```
Comment (mikewiebe) [inline plugins/action/common/prepare_plugins/prep_107_policy.py:45]: I don't see where you are making any modifications to the extended model_data before you pass it back to the ansible playbook.  Not sure I follow how this is being used.

File: roles/dtc/common/tasks/ndfc_policy.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# SPDX-License-Identifier: MIT
+
+---
+
```
Comment (mikewiebe) [inline roles/dtc/common/tasks/ndfc_policy.yml:23]: Reminder to add a changes detected flag for the policy stuff.  Example from fabric section

https://github.com/netascode/ansible-dc-vxlan/blob/develop/roles/dtc/common/tasks/ndfc_fabric.yml#L24

File: plugins/action/common/prepare_plugins/prep_107_policy.py (hunk 1 above)
Comment (mtarking) [reply plugins/action/common/prepare_plugins/prep_107_policy.py:45]: It should be just a mutable object reference getting updated if not mistaken.

The data model looks like this:

```
---
vxlan:
  policy:
    <snip>
    switches:
      - name: 10.15.33.13
        groups:
          - server leafs
      - name: netascode-leaf4
        groups:
          - server leafs
```

Post the prepare plugin, the data looks like this:

```
{
    "vxlan": {
        "policy": {
            <snip>
            "switches": [
                {
                    "groups": [
                        "server leafs"
                    ],
                    "name": "10.15.33.13"
                },
                {
                    "groups": [
                        "server leafs"
                    ],
                    "name": "10.15.33.16"
                }
            ]
        }
    }
}
```

I can try to do this differently if you think it's needed.

File: roles/dtc/common/tasks/ndfc_policy.yml (hunk 2 above)
Comment (mtarking) [reply roles/dtc/common/tasks/ndfc_policy.yml:23]: @mikewiebe, updated and dropped you a msg in Webex.

File: plugins/action/common/prepare_plugins/prep_107_policy.py (hunk 1 above)
Comment (mikewiebe) [reply plugins/action/common/prepare_plugins/prep_107_policy.py:42]: Chatted about this and I am fine with the changes

File: roles/dtc/common/tasks/ndfc_policy.yml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+- name: Set File Change Flag Based on File Diff Result
+  ansible.builtin.set_fact:
+    changes_detected_policy: true
+  delegate_to: localhost
+  when:
+    - file_diff_result.file_data_changed
+    - check_roles['save_previous']
+
+#########################################
```
Comment (mikewiebe) [inline roles/dtc/common/tasks/ndfc_policy.yml:80]: Remove these old commented out tasks before commit.

File: roles/dtc/common/templates/ndfc_policy.j2 [hunk 4]
```diff
@@ -0,0 +1,42 @@
+---
+# This NDFC policy and switch attachments config data structure is auto-generated
+# DO NOT EDIT MANUALLY
+#
+{# {% set switches = MD_Extended.vxlan.policy.switches | community.general.json_query('[*].name') | flatten | unique | sort %} #}
```
Comment (mikewiebe) [inline roles/dtc/common/templates/ndfc_policy.j2:5]: Can you remove this line 5?

File: roles/validate/files/rules/required_rules/501_policy_cross_reference.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            if inventory["vxlan"].get("policy", None):
+                if inventory["vxlan"].get("policy").get("policies", None):
+                    policies = inventory["vxlan"]["policy"]["policies"]
+                    for policy in policies:
+                        filename = policy.get("filename", None)
+
+                        if ((filename and policy.get("template_name", None) and policy.get("template_vars", None)) or
+                                (filename and policy.get("template_vars", None))):
+                            results.append(
+                                "Policy definitions are filename with .config or .cfg that defaults template_name to NDFC freeform or "
```
Comment (mikewiebe) [inline roles/validate/files/rules/required_rules/501_policy_cross_reference.py:23]: This result message is a bit confusing to me in the way it reads.

File: roles/dtc/common/templates/ndfc_policy.j2 (hunk 4 above)
Comment (mtarking) [reply roles/dtc/common/templates/ndfc_policy.j2:5]: Done

File: roles/dtc/common/tasks/ndfc_policy.yml (hunk 3 above)
Comment (mtarking) [reply roles/dtc/common/tasks/ndfc_policy.yml:80]: Done

File: roles/validate/files/rules/required_rules/501_policy_cross_reference.py (hunk 5 above)
Comment (mtarking) [reply roles/validate/files/rules/required_rules/501_policy_cross_reference.py]: Reworded result msg.

**Review by mikewiebe — APPROVED** [review]

**Comment by mtarking** [comment]
Resolves #72

**Pull request merged**